dag_in_context/src/dag2svg.rs

```diff
@@ -76,22 +129,50 @@ impl Expr {
 
     fn to_dot_with(self: &RcExpr, conv: &mut DotConverter) -> Vec<Stmt> {
         let id = Rc::as_ptr(self);
-        if !conv.done.insert(id) {
+        if !conv.done.insert(UniqueExpr::new(conv.current_scope, self)) {
             return vec![];
         }
         match self.as_ref() {
             Expr::DoWhile(inputs, body) => {
                 let mut stmts = inputs.to_dot_with(conv);
+                let vertex_outside = conv.graphviz_vertex(self);
+                let id_outside = conv.graphviz_id(self);
                 stmts.push(Stmt::Edge(Edge {
-                    ty: EdgeTy::Pair(conv.graphviz_vertex(self), conv.graphviz_vertex(inputs)),
+                    ty: EdgeTy::Pair(vertex_outside.clone(), conv.graphviz_vertex(inputs)),
                     attributes: vec![],
                 }));
+
+                let scope_before = conv.current_scope;
+                conv.current_scope = id;
                 stmts.push(Stmt::Subgraph(Subgraph {
                     stmts: body.to_dot_with(conv),
-                    id: conv.graphviz_id(self),
+                    id: id_outside,
+                }));
+                stmts.push(Stmt::Edge(Edge {
+                    ty: EdgeTy::Pair(vertex_outside, conv.graphviz_vertex(body)),
+                    attributes: vec![],
                 }));
+                // restore scope of the dowhile
+                conv.current_scope = scope_before;
```

Comment:
Why is this in `DoWhile` but not below in `Function`?

Reply:
`Function` is a top-level scope so there's no scope above it

Reply:
added comment